cadcutils/cadcutils/net/ws.py

```diff
@@ -461,36 +480,176 @@ def upload_file(self, url, src, md5_checksum=None, **kwargs):
         if not headers:
             kwargs['headers'] = headers
         headers[HTTP_LENGTH] = str(stat_info.st_size)
-        if md5_checksum:
-            net.add_md5_header(headers=headers, md5_checksum=md5_checksum)
-        else:
-            if stat_info.st_size <= MAX_MD5_COMPUTE_SIZE:
-                hash_md5 = hashlib.md5()
-                with open(src, "rb") as f:
-                    for chunk in iter(lambda: f.read(4096), b""):
-                        hash_md5.update(chunk)
-                md5 = hash_md5.hexdigest()
-                net.add_md5_header(headers=headers, md5_checksum=md5)
-
-        with util.Md5File(src, 'rb') as reader:
-            response = self._get_session().put(
-                url,
-                data=reader,
-                verify=self.verify, **kwargs)
-        if md5_checksum and (md5_checksum != reader.md5_checksum):
-            raise exceptions.TransferException(
-                'Provided checksum for {} does not match checksum'
-                'of bytes read: {} != {}'.format(src, md5_checksum,
-                                                 reader.md5_checksum))
-        # check the file made it OK
-        dest_md5 = net.extract_md5(response.headers)
-        if dest_md5 != reader.md5_checksum:
-            raise exceptions.TransferException(
-                'File {} not properly uploaded. Mistmatched md5 src vs '
-                'dest: {} vs {}'.format(src, reader.md5_checksum, dest_md5)
-            )
-
-    def _resolve_destination_file(self, dest, src_md5, default_file_name):
+        src_md5 = md5_checksum
+        if not src_md5 and stat_info.st_size <= MAX_MD5_COMPUTE_SIZE:
+            hash_md5 = hashlib.md5()
+            with open(src, "rb") as f:
+                for chunk in iter(lambda: f.read(4096), b""):
+                    hash_md5.update(chunk)
+            src_md5 = hash_md5.hexdigest()
+
+        if src_md5:
+            net.add_md5_header(headers=headers, md5_checksum=src_md5)
+            if stat_info.st_size < FILE_SEGMENT_THRESHOLD:
+                # no transactions needed.
+                with open(src, 'rb') as reader:
+                    response = self._get_session().put(
+                        url,
+                        data=reader,
+                        verify=self.verify, **kwargs)
+                self.logger.debug('{} uploaded (HTTP {})'.format(
+                    src, response.status_code))
```

Comment:
No md5sum check from the response, for this case? Should that be reflected in the docs, since it's a change in behaviour? 

No retries for this case?

Reply:
Sending the pre-computed md5 checksum in the request allows the server to check for mismatch and fail if necessary. Successful return means that they match.

Good point about retries - I've added them.